mne/io/base.py

```diff
@@ -836,6 +836,102 @@ def __setitem__(self, item, value):
+            Copy of the data in the given range.
+        times : ndarray, shape (n_times,)
+            Times associated with the data samples. Only returned if
+            return_times=True.
+
+        Notes
+        -----
+
+        .. versionadded:: 0.14.0
+        """
+        if picks is None:
+            picks = range(self.info['nchan'])
+        if self.annotations is None or reject_by_annotation is None:
+            data, times = self[picks, start:stop]
+            if return_times:
+                return data, times
+            return data
+        if reject_by_annotation not in ['omit', 'NaN']:
+            raise ValueError("reject_by_annotation must be None, 'omit' or "
+                             "'NaN'. Got %s." % reject_by_annotation)
+        sfreq = self.info['sfreq']
+        bads = [idx for idx, desc in enumerate(self.annotations.description)
+                if desc.upper().startswith('BAD')]
+        onsets = self.annotations.onset[bads]
+        onsets = _sync_onset(self, onsets)
+        ends = onsets + self.annotations.duration[bads]
+        omit = np.concatenate([np.where(onsets > stop / sfreq)[0],
+                               np.where(ends < start / sfreq)[0]])
+        onsets, ends = np.delete(onsets, omit), np.delete(ends, omit)
+        if len(onsets) == 0:
+            data, times = self[picks, start:stop]
+            if return_times:
+                return data, times
+            return data
+        stop = min(stop, self.n_times)
+        order = np.argsort(onsets)
+        onsets = self.time_as_index(onsets[order])
+        ends = self.time_as_index(ends[order])
+
+        np.clip(onsets, start, stop, onsets)
+        np.clip(ends, start, stop, ends)
+        used = np.ones(stop - start, bool)
+        for onset, end in zip(onsets, ends):
+            if onset >= end:
+                continue
+            used[onset - start: end - start] = False
+        used = np.concatenate([[False], used, [False]])
+        starts = np.where(~used[:-1] & used[1:])[0] + start
+        stops = np.where(used[:-1] & ~used[1:])[0] + start
+        if reject_by_annotation == 'omit':
+
+            data = np.zeros((len(picks), np.sum([b - a for a, b in
+                                                 zip(starts, stops)])))
+            times = np.zeros(data.shape[1])
+            idx = 0
+            for start, stop in zip(starts, stops):  # get the data
+                if start == stop:
+                    continue
+                end = idx + stop - start
+                data[:, idx:end], times[idx:end] = self[picks, start:stop]
+                idx = end
+        else:
+            times = self.times[start:stop].copy()
+            data = np.zeros((len(picks), len(times)))
+            data.fill(np.nan)  # cannot be inserted to existing data array
+            for start, stop in zip(starts, stops):
+                data[:, start:stop] = self[picks, start:stop][0]
```

Comment:
I thought I tried doing something similar and it failed to set it to NaN for some reason. Seems to work now.